[Tech Docs] Libcurl transport adapter (#3484)
Azure/azure-sdk-for-cpp: doc/LibcurlTransportAdapter.md

# Azure SDK for C++ - Libcurl Transport Adapter

## Azure SDK for C++

The [Azure SDK for C++](https://github.com/Azure/azure-sdk-for-cpp) enables an easy and consistent experience for consuming Azure services.
Azure services offer customers a simple way to perform many kinds of operations online. As such, the operations work as a client/server communication where the operations are requested by a client (customer) and executed by an Azure service over a network.

### HTTP

Most Azure services use the HTTP protocol for client/service communication. The protocol defines the structure of how a request and a response must look like.
 
Modern programming languages like C#, Java, and JS have an HTTP client as part of the language itself.  For these languages, the runtime library provides classes making HTTP easy to use.
 

> However, for C++, the runtime library does not include any HTTP classes; you either need to write an HTTP implementation yourself or you could use some third-party library.  Keep in mind, that you also have to consider the OSes (Windows, Linux, Mac, RTOS, etc.) you plan to run your application on. Libcurl, for example, is one of the most popular cross-OS libraries.

Reviewer: Our C++ SDK doesn't support RTOSes - I'd remove that

### Easy handle

Let's start by talking about libcurl fundamentals. Libcurl is written in C, as such it is fast and practical. It provides two main modes of operation, the easy handle and the multi handle. The easy handle is the starting point for both. The multi handle is nothing but a group of easy handles.
 
An easy handle represents a network request to be performed. For a developer, the typical implementation flow is:

1.	Create an easy handle (as a raw pointer).
2.	Ask libcurl to init the handle. Libcurl handles memory allocation.
3.	Configure the handle according to what needs to be requested to service.
4.	Tell libcurl to take the easy handle and perform the request.
 
As simple as that, libcurl hides a ton of complexity from you. Libcurl asks the Operating System for a network socket and establishes a connection. It even takes care of securing the connection with TLS when requested. It sends the HTTP request to the service, and it receives the HTTP response. It sounds magical! However, it is not enough for Azure SDK's requirements.

### Azure SDK requirements

> From all the operations supported by Azure services, there's one that can't be fully completed by following the libcurl typical flow mentioned before. It is the download operation from the Azure Storage service which requires something else. When the size of the downloaded item is bigger than the amount of memory in the system, for example, downloading 1Tb from an Android device with only 8GB of memory, or from a Windows PC with 32 GB of memory, there should be no complications for an SDK client. However, based on the typical libcurl flow, during step 4, libcurl blocks the application until all the bytes from the server have been received, and that's not acceptable from the SDK perspective.

Reviewer: I think you mean downloading TO (not from) an Android device.

# Azure SDK for C++ - Libcurl Transport Adapter

## Azure SDK for C++

> The [azure SDK for C++](https://github.com/Azure/azure-sdk-for-cpp) enables and easy and consistent experience for consuming Azure services.

Reviewer 2: ```suggestion
The [azure SDK for C++](https://github.com/Azure/azure-sdk-for-cpp) enables an easy and consistent experience for consuming Azure services.
```

> The [azure SDK for C++](https://github.com/Azure/azure-sdk-for-cpp) enables and easy and consistent experience for consuming Azure services.

Reviewer 3: ```suggestion
The [Azure SDK for C++](https://github.com/Azure/azure-sdk-for-cpp) enables and easy and consistent experience for consuming Azure services.
```

> Azure services offer customers a simple way to perform many kinds of operations online. As such, the operations work as a client - server communication where the operations are requested by a client (customer) and executed by a server (Azure server). Communication goes from one point to another, typically through a network.

Reviewer: ```suggestion
Azure services offer customers a simple way to perform many kinds of operations online. As such, the operations work as a client/ erver communication where the operations are requested by a client (customer) and executed by an Azure service over a network.
```

Reviewer 4: Nit: Typo in client/server (written as client/erver).

### HTTP

> Many Azure services use TCP/HTTP protocol to exchange messages with a client. The protocol defines the structure of how a request and a response must look like, so, while Server and a Client both follow the protocol rules, they can establish a channel to interchange messages.

Reviewer: ```suggestion
Most Azure services use the HTTP protocol for client/service communication. The protocol defines the structure of how a request and a response must look like.
```

> Modern programing languages like C#, Java or JS have an HTTP client as part of the language itself.   As a C# developer, I don't need to learn about the specifics of the protocol, the language has already abstracted it all, and all that I need are a few classes to instantiate.

Reviewer: ```suggestion
Modern programming languages like C#, Java, and JS have an HTTP client as part of the language itself.  For these languages, the runtime library provides classes making HTTP easy to use.
```

> For the case of C++, the language does not include any sort of abstraction for HTTP protocol. As a C++ developer, I would either need to write an HTTP implementation for my application, which would depend on the system (Windows, Linux, Mac, RTOS, etc.), or, I could use some third party library which would already provide a multi-platform implementation for the protocol.  Libcurl, for example, is one of the most popular libraries.

Reviewer: ```suggestion
However, for C++, the runtime library does not include any HTTP classes; you either need to write an HTTP implementation yourself or you could use some third-party library.  Keep in mind, that you also have to consider the OSes (Windows, Linux, Mac, RTOS, etc.) you plan to run your application on. Libcurl, for example, is one of the most popular cross-OS libraries.
```

### Replaceable HTTP Transport

> One of the most interesting features of the Azure SDK for C++ is how, as a user, I can take control and choose the HTTP implementation to be used to communicate to Azure. That implementation is known by the SDK as an HTTP transport adapter. The azure-core library includes an HTTP namespace where it has defined classes to abstract the HTTP components (request, response, send, etc.). While the Request and the Response are well defined based on the HTTP protocol [RFC](https://datatracker.ietf.org/doc/html/rfc7230), the send operation is left as a virtual method in the HttpTransport abstract class. Then, a transport adapter is a derived class that implements the send operation. Learn more about HTTP Transport [here](https://github.com/Azure/azure-sdk-for-cpp/blob/main/doc/HttpTransportAdapter.md#http-transport-adapter).

Reviewer 4: ```suggestion
One of the most interesting features of the Azure SDK for C++ is that a customer has the ability to choose the HTTP implementation to be used to communicate with Azure services. That implementation is known by the SDK as an HTTP transport adapter. The azure-core library includes an HTTP namespace where it has defined classes to abstract the HTTP components (request, response, send, etc.). While the Request and the Response are well defined based on the HTTP protocol [RFC](https://datatracker.ietf.org/doc/html/rfc7230), the send operation is left as a virtual method in the HttpTransport abstract class. Then, a transport adapter is a derived class that implements the send operation. Learn more about HTTP Transport [here](https://github.com/Azure/azure-sdk-for-cpp/blob/main/doc/HttpTransportAdapter.md#http-transport-adapter).
```

> One of the most interesting features of the Azure SDK for C++ is how, as a user, I can take control and choose the HTTP implementation to be used to communicate to Azure. That implementation is known by the SDK as an HTTP transport adapter. The azure-core library includes an HTTP namespace where it has defined classes to abstract the HTTP components (request, response, send, etc.). While the Request and the Response are well defined based on the HTTP protocol [RFC](https://datatracker.ietf.org/doc/html/rfc7230), the send operation is left as a virtual method in the HttpTransport abstract class. Then, a transport adapter is a derived class that implements the send operation. Learn more about HTTP Transport [here](https://github.com/Azure/azure-sdk-for-cpp/blob/main/doc/HttpTransportAdapter.md#http-transport-adapter).

Reviewer: ```suggestion
One of the more interesting features of the Azure SDK for C++ is how, as a user, you can take control and choose the HTTP implementation to be used to communicate to Azure. That implementation is known by the SDK as an HTTP _transport adapter_. The azure-core library includes an HTTP namespace where it defines classes to abstract the HTTP components (request, response, send, etc.). While the Request and the Response are well defined based on the HTTP protocol [RFC](https://datatracker.ietf.org/doc/html/rfc7230), the send operation is left as a virtual method in the HttpTransport abstract class. Then, a transport adapter is a derived class that implements the Send operation. Learn more about HTTP transport adapter [here](https://github.com/Azure/azure-sdk-for-cpp/blob/main/doc/HttpTransportAdapter.md#http-transport-adapter).
```

## Libcurl Transport Adapter

> The azure-core-cpp library provides a transport adapter implemented with libcurl. The next paragraphs will mention some of the limitations of libcurl and the code that is part of the LibcurlTransportAdatper  (or LTA) that satisfy the expectations for an SDK client.

Reviewer: ```suggestion
The azure-core-cpp library provides a transport adapter implemented with libcurl. The next paragraphs mentions some of the limitations of libcurl and the code that is part of the LibcurlTransportAdatper  (LTA) that satisfy the expectations for an SDK client.
```

### Easy handle

Let's start by talking about libcurl fundamentals. Libcurl is written in C, as such it is fast and practical. It provides two main modes of operation, the easy handle and the multi handle. The easy handle is the starting point for both. The multi handle is nothing but a group of easy handles.
 
An easy handle represents a network request to be performed. For a developer, the typical implementation flow is:

1.	Create an easy handle (as a raw pointer).
2.	Ask libcurl to init the handle. Libcurl handles memory allocation.

> 3.	Configure handle according to what needs to be requested to server.

Reviewer: ```suggestion
3.	Configure the handle according to what needs to be requested to service.
```

4.	Tell libcurl to take the easy handle and perform the request.
 

> As simple as that, libcurl is putting away a ton of complexity from developer's hands. If step 3 was done correctly, libcurl would ask the Operating System for a network socket to speak to the network. It will establish a connection. It will even take care of securing the connection with TLS when requested. It will talk to a server following HTTP spec and it will receive and parse a response from the network. It sounds magical! However, it is not enough for the Azure SDK requirements.

Reviewer: ```suggestion
As simple as that, libcurl hides a ton of complexity from you. Libcurl asks the Operating System for a network socket and establishes a connection. It even takes care of securing the connection with TLS when requested. It sends the HTTP request to the service, and it receives the HTTP response. It sounds magical! However, it is not enough for Azure SDK's requirements.
```

### Azure SDK requirements

> From all the operations supported by Azure services, there's one that can't be fully completed by following the libcurl typical flow mentioned before. It is the download operation from the Azure Storage service which requires something else. When the size of the downloaded item is bigger than the amount of memory in the system, for example, downloading 1Tb from an Android device with only 8GB of memory, or from a Windows PC with 32 GB of memory, there should be no complications for an SDK client. However, based on the typical libcurl flow, during step 4, libcurl blocks the application until all the bytes from the server have been received, and that's not acceptable from the SDK perspective.

Reviewer: ```suggestion
From all the operations supported by Azure services, there's one that can't be fully completed by following the libcurl typical flow mentioned before. An upload or download operation may require the ability to send/receive significant amounts of data. For example, an Azure Storage blob can be well over a Terabyte in size. However, based on the typical libcurl flow, during step 4, libcurl blocks the application until all the bytes from the service have been received, and that's not acceptable from the SDK perspective.
```

Author: It is only `Download` the operation which can't be done with libcurl proper way.
It is not because of the size, but because we want customers to decide when to read bytes from network and how much

#### Stream response on demand

> The expectation when using the SDK client to download bytes is to enable an application to read any amount of data from the network, on demand. An application must be able to read any amount of data from the network, then, it must be able to do something with that data and decide when to read some more. 

Reviewer 4: Too many ',' (commas). Possibly "An application must be able to read any amount of data from the network, do something with that data and then decide when to read some more" (or something like that).

> The expectation when using the SDK client to download bytes is to enable an application to read any amount of data from the network, on demand. An application must be able to read any amount of data from the network, then, it must be able to do something with that data and decide when to read some more. 

Reviewer: ```suggestion
To enable the processing of large amounts of data, the Azure SDK offers a stream abstraction over the sending/receiving of HTTP request/response bodies. ```

 
A simple example to visualize this requirement is to think about a big download (say 1Tb) as a media file with a high quality (say a 4K movie). Watching a few seconds of the movie, from any device, should not require downloading the entire movie. Downloading a few seconds would be equivalent to reading a few hundred bytes, which are played from the device while some more seconds (data) are downloaded. Whatever is already watched can be overridden by some more content.

> #### No intermediate buffer

Reviewer 4: Is this section necessary? It describes state that no longer applies to the Azure SDK. Maybe just include the "no intermediate buffer" as a requirement from the previous section?

Author: `Stream response on demand` and `No intermediate buffer` are subtitles from `Azure SDK requirements`.
Each one is a requirement

Older versions of the Azure SDK have worked around the libcurl programing model by using an intermediate buffer to fetch the entire response from the server and let the program continue from step 4 (from the steps mentioned above). By doing this, the maximum download size gets limited to what the system memory can handle. Older SDK versions would support 2GB as the maximum download size.
 
Another approach is to be able to request data from the server starting at some specific offset. Then, downloading big content would be equivalent to sending one request after another, asking for data starting on the last downloaded data. However, this strategy adds unnecessary extra messaging.  Each request is parsed by the server to produce a response that is sent back to the client. The client also parses the response. These server and client parsing would be happening one after each other, so it is not efficient.

> ### Poor man's substitute

Reviewer 4: What is a "Poor man's substitute"? 

Also, consider rewording this heading because it is somewhat derogatory towards poor people.

 
1.	Resolve DNS from request URL.
2.	Get a network socket from the OS.
3.	Establish TCP connection to the server.
4.	Secure connection with TLS when using HTTPS.
5.	Create an HTTP request from the handle and write it to the network socket.
6.	Wait for socket to be readable (means server has responded) and get an HTTP status line and headers.
7. If there's a payload body (data to be downloaded), a loop is started here as:
*  7.1 Read data from the network socket.
*  7.2 Invoke the application callback passing the data read from the socket.
* 7.3 Repeat until the entire response has been received.
 

> This flow demonstrates how using libcurl to download data requires developers to use a callback pattern. For the example of reproducing a media file (mentioned above), playing media must happen as part of the callback that is invoked by libcurl, which is very unfortunate. 

Reviewer 4: It's more than just "unfortunate" - it's untenable because of the requirements associated with multimedia rendering.

 
There are valid alternatives for applications, but it might be complex approaches. For example, using a multi-threading strategy, an application can call curl_easy_perform from one thread and use the read callback to copy the data from the response into a memory buffer. Then, another thread can play the media from the buffer. This strategy requires threads synchronization and complicates the debugging experience.
 

> Some other designs might help applications to deal with this. But it is a different story for libraries.  The Azure SDK is a library, and it ensures a consistent experience regardless of the programing language, as such, the API is not shaped by its implementation details. The libcurl implementation details from the Azure SDK is considered a poor man's substitute by [libcurl author](https://github.com/curl/curl/issues/6009#issuecomment-698611402), and it is explained next.

Reviewer 4: I'm not sure this paragraph is necessary.

Author: It is the explanation of why the approach is called `a poor man substitute`.  As that's the name given by libcurl author.
But, yes, I think I know what you mean.  I was trying to add some fun/drama around it;  but it is not adding any real technical value >D

Reviewer 4: Drama is wonderful when telling stories (arguably it's essential). But it's not necessarily a great solution in a technical document :(.

FWIW, the phrase "Poor Man's" does not appear in the policheck database, nor did any online definitions I found indicate it was considered inappropriate. But it is a term I wouldn't necessarily use.

 
Most of the libcurl handle configuration becomes useless when the handle is changed to custom protocol. For example, it becomes irrelevant creating a headers list, or setting an HTTP method (GET, POST, etc.). Libcurl will ignore it, and it will be now the customer who takes connection ownership.  What this means for the libcurlTransportAdapter is that it has to learn how to speak raw HTTP and translate an Azure:: Core:: Http:: Request into a pure HTTP call.

#### Receiving raw HTTP

The most tedious task, while speaking raw HTTP, is reading and parsing responses. As mentioned before, the main purpose of using the poor man's substitute (custom protocol) to re-implement HTTP (even though libcurl provides HTTP support natively) is to enable an Azure SDK client to start a download operation and let customers to decide when to pull chunks of data from the response (directly from the network socket, without any intermediate buffers or using callback functions).
 

> An HTTP request is nothing but a string with a specific format, it has special text delimiters to distinguish between the request line, headers and body. As soon as the server receives and parses the request, it would trigger an operation and eventually produce a response. The response needs to be HTTP formatted using text delimiters to indicate where the headers section starts and where it ends. Then the server will start sending small parts of it. On the client side, calling _curl_easy_recv()_ will get any bytes that have arrived. As soon as the function pulls the bytes from the socket, the socket will receive more data that will be written on top of the previous data. The function _curl_easy_recv()_ is a way to tell the OS, _“Hey, I got this data, you can write on top of it now"_.

Reviewer 4: One could argue that the same applies to ANY text based protocol (POP3, IMAP4, NNTP, FTP, SMTP, MIME, BASE64, PEM, etc) - they are all strings with a specific format.
